aiodataloader.py

```diff
@@ -1,29 +1,82 @@
-from asyncio import gather, ensure_future, get_event_loop, iscoroutine, iscoroutinefunction
-from collections import namedtuple
-from collections.abc import Iterable
+from asyncio import (
+    AbstractEventLoop,
+    Future,
+    gather,
+    ensure_future,
+    get_event_loop,
+    iscoroutine,
+    iscoroutinefunction,
+)
 from functools import partial
 
-from typing import List  # flake8: noqa
+from typing import (
+    Any,
+    Callable,
+    Coroutine,
+    Dict,
+    Generic,
+    Iterable,
+    Iterator,
+    List,
+    Optional,
+    TypeVar,
+    Union,
+)
+from typing_extensions import Protocol, TypeGuard
+
+__version__ = "0.2.1"
+
+
+def iscoroutinefunctionorpartial(fn: Callable) -> TypeGuard[Callable[..., Coroutine]]:
+    return iscoroutinefunction(fn.func if isinstance(fn, partial) else fn)
 
 
-__version__ = '0.2.1'
+KeyT = TypeVar("KeyT")
+ReturnT = TypeVar("ReturnT")
+CacheKeyT = TypeVar("CacheKeyT")
+T = TypeVar("T")
 
-Loader = namedtuple('Loader', 'key,future')
 
+class BatchLoadFnProto(Protocol[KeyT, ReturnT]):
+    async def __call__(self, keys: List[KeyT]) -> List[ReturnT]:
+        ...
 
-def iscoroutinefunctionorpartial(fn):
-    return iscoroutinefunction(fn.func if isinstance(fn, partial) else fn)
 
+class Loader(tuple, Generic[KeyT, ReturnT]):
+    # NamedTuple has a metaclass conflict with Generic that prevents
+    # it from being used generically, and dataclasses aren't supported until
+    # Python 3.7
 
-class DataLoader(object):
+    key: KeyT
+    future: "Future[ReturnT]"
+
+    def __new__(cls, key: KeyT, future: "Future[ReturnT]"):
+        self = tuple.__new__(cls, (key, future))
+        self.key = key
+        self.future = future
+        return self
 
-    batch = True
-    max_batch_size = None  # type: int
-    cache = True
+    def __repr__(self) -> str:
+        return f"Loader(key={self.key}, future={self.future})"
 
-    def __init__(self, batch_load_fn=None, batch=None, max_batch_size=None,
-                 cache=None, get_cache_key=None, cache_map=None, loop=None):
 
+class DataLoader(Generic[KeyT, ReturnT]):
+
+    batch: bool = True
+    max_batch_size: Optional[int] = None
+    cache: Optional[bool] = True
+    _queue: List[Loader[KeyT, ReturnT]]
+
+    def __init__(
+        self,
+        batch_load_fn: Optional[BatchLoadFnProto[KeyT, ReturnT]] = None,
+        batch: Optional[bool] = None,
+        max_batch_size: Optional[int] = None,
+        cache: Optional[bool] = None,
+        get_cache_key: Optional[Callable[[KeyT], CacheKeyT]] = None,
```

Comment:
I think this ought to be `Optional[Callable[[KeyT], Union[CacheKeyT, KeyT]]]`, otherwise a type error occurs on this:
```
self.get_cache_key = get_cache_key or (lambda x: x)
```

Reply:
Yeah that tracks, I'll do a thing